Write operations fail when talking to Synology NAS over SFTP

## Output of `restic version`

Today's tip of trunk

```
$ ./restic version
debug enabled
restic 0.2.0 (v0.2.0-176-g2635152)
compiled at 2016-08-28 16:09:23 with go1.6
```
## Expected behavior

This is a follow up to issue #592 trying to get SFTP happy with the various errors from talking to my weird Synology NAS.

Basic read operations worked, but they had a problem with locking so I guess it was a problem writing.

So here I tried to create a new test repository.
## Actual behavior

```
$ DEBUG_LOG=/tmp/restic-debug.log ./restic -r sftp:p4:test1 init
debug log file /tmp/restic-debug.log
debug enabled
panic: runtime error: index out of range

goroutine 1 [running]:
panic(0x9d8480, 0xc820010150)
    /usr/local/go/src/runtime/panic.go:464 +0x3e6
restic/debug.Log(0xaaa2d0, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
    /tmp/restic-build-165477600/src/restic/debug/debug.go:149 +0xa19
restic/backend/sftp.(*SFTP).Close(0xc820131110, 0x0, 0x0)
    /tmp/restic-build-165477600/src/restic/backend/sftp/sftp.go:526 +0x6a
restic/backend/sftp.Create(0x7ffd7680f490, 0x5, 0xa95640, 0x3, 0xc82000af80, 0x3, 0x4, 0x3, 0x0, 0x0)
    /tmp/restic-build-165477600/src/restic/backend/sftp/sftp.go:182 +0x511
restic/backend/sftp.CreateWithConfig(0x0, 0x0, 0x7ffd7680f48d, 0x2, 0x7ffd7680f490, 0x5, 0x1, 0x0, 0x0)
    /tmp/restic-build-165477600/src/restic/backend/sftp/sftp.go:195 +0x222
main.create(0x7ffd7680f488, 0xd, 0x0, 0x0, 0x0, 0x0)
    /tmp/restic-build-165477600/src/cmds/restic/global.go:320 +0x9ae
main.CmdInit.Execute(0xd87960, 0xc820130870, 0x0, 0x3, 0x0, 0x0)
    /tmp/restic-build-165477600/src/cmds/restic/cmd_init.go:18 +0x169
main.(*CmdInit).Execute(0xc82002a088, 0xc820130870, 0x0, 0x3, 0x0, 0x0)
    <autogenerated>:32 +0xb9
github.com/jessevdk/go-flags.(*Parser).ParseArgs(0xc820014960, 0xc82000a3d0, 0x3, 0x3, 0x0, 0x0, 0x0, 0x0, 0x0)
    /tmp/restic-build-165477600/src/github.com/jessevdk/go-flags/parser.go:278 +0x8dd
github.com/jessevdk/go-flags.(*Parser).Parse(0xc820014960, 0x0, 0x0, 0x0, 0x0, 0x0)
    /tmp/restic-build-165477600/src/github.com/jessevdk/go-flags/parser.go:154 +0x9b
main.main()
    /tmp/restic-build-165477600/src/cmds/restic/main.go:32 +0x1ca
```

debug log:

```
2016/08/28 16:19:50 0.000 [    restic]   1 main.go:30 main []string{"./restic", "-r", "sftp:p4:test1", "init"}
2016/08/28 16:19:50 0.000 [      open]   1 global.go:308 parsing location sftp:p4:test1
2016/08/28 16:19:50 0.000 [      open]   1 global.go:319 create sftp repository at sftp.Config{User:"", Host:"p4", Dir:"test1"}
2016/08/28 16:19:50 0.000 [sftp.CreateWithConfig]   1 sftp.go:194 config { p4 test1}
2016/08/28 16:19:50 0.000 [          sftp.Create]   1 sftp.go:162 ssh [p4 -s sftp]
```

The directory does appear to exist on the disk with reasonable permissions, but it never asked me about a password.

```
wscott@p4:~$ ls -l test1
total 24
drwx------ 2 wscott users 4096 Aug 28 16:19 data
drwx------ 2 wscott users 4096 Aug 28 16:19 index
drwx------ 2 wscott users 4096 Aug 28 16:19 keys
drwx------ 2 wscott users 4096 Aug 28 16:19 locks
drwx------ 2 wscott users 4096 Aug 28 16:19 snapshots
drwx------ 2 wscott users 4096 Aug 28 16:19 tmp
wscott@p4:~$ ls -lR test1
test1:
total 24
drwx------ 2 wscott users 4096 Aug 28 16:19 data
drwx------ 2 wscott users 4096 Aug 28 16:19 index
drwx------ 2 wscott users 4096 Aug 28 16:19 keys
drwx------ 2 wscott users 4096 Aug 28 16:19 locks
drwx------ 2 wscott users 4096 Aug 28 16:19 snapshots
drwx------ 2 wscott users 4096 Aug 28 16:19 tmp

test1/data:
total 0

test1/index:
total 0

test1/keys:
total 0

test1/locks:
total 0

test1/snapshots:
total 0

test1/tmp:
total 0
```

It might again be the Synology to blame but again we are missing some error handling.

The code in #597 did fix the initial problem.

And yes from linux, 'sftp p4' just works and allows me to read and write files without a password.

Here is a new example and shows the same problem as saw in the other issue.

Repostory create looks fine
```
$ DEBUG_LOG=/tmp/restic-debug.log ./restic -r sftp:p4:test2 init
debug log file /tmp/restic-debug.log
debug enabled
enter password for new backend: 
enter password again: 
created restic backend 48690db4a4 at sftp:p4:test2

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
```

And a backup works, but fails at the end. I suspect something related to locking.
```
$ DEBUG_LOG=/tmp/restic-debug.log ./restic -r sftp:p4:test1 backup /home/wscott/src/restic/README.md 
debug log file /tmp/restic-debug.log
debug enabled
enter password for repository: 
scan [/home/wscott/src/restic/README.md]
scanned 0 directories, 1 files in 0:00
[0:00] 100.00%  0B/s  3.536 KiB / 3.536 KiB  1 / 1 items  0 errors  ETA 0:00 
duration: 0:00, 0.04MiB/s
snapshot e1239daa saved
error in cleanup handler: file does not exist
```

And here is the debug log:
```
2016/08/29 07:16:39 0.000 [    restic]   1 main.go:30 main []string{"./restic", "-r", "sftp:p4:test2", "init"}
2016/08/29 07:16:39 0.000 [      open]   1 global.go:308 parsing location sftp:p4:test2
2016/08/29 07:16:39 0.000 [      open]   1 global.go:319 create sftp repository at sftp.Config{User:"", Host:"p4", Dir:"test2"}
2016/08/29 07:16:39 0.000 [sftp.CreateWithConfig]   1 sftp.go:194 config { p4 test2}
2016/08/29 07:16:39 0.000 [          sftp.Create]   1 sftp.go:162 ssh [p4 -s sftp]
2016/08/29 07:16:40 0.000 [           sftp.Close]   1 sftp.go:526 
2016/08/29 07:16:40 0.062 [           sftp.Close]   1 sftp.go:532 Close returned error <nil>
2016/08/29 07:16:40 0.000 [            sftp.Wait]  11 sftp.go:75 ssh command exited, err <nil>
2016/08/29 07:16:40 0.000 [            sftp.Open]   1 sftp.go:118 open backend with program ssh, [p4 -s sftp] at test2
2016/08/29 07:16:44 0.000 [            sftp.Test]   1 sftp.go:426 type config, name 
2016/08/29 07:16:44 0.000 [    Repo.CreateConfig]   1 config.go:63 New config: repository.Config{Version:0x1, ID:"48690db4a425a5b92150300e5bec8ae68f4df9815bf659fb020c989378bbb1dc", ChunkerPolynomial:0x34e234ffc67b57}
2016/08/29 07:16:46 0.000 [    repository.AddKey]   1 key.go:169 calibrated KDF parameters are {262144 1 2}
2016/08/29 07:16:46 0.000 [            sftp.Save]   1 sftp.go:365 save <key/5f9df25253> bytes to 447
2016/08/29 07:16:46 0.002 [            sftp.Save]   1 sftp.go:379 save <key/5f9df25253> (447 bytes) to test2/tmp/temp-a956b9862f2715b6a8e6
2016/08/29 07:16:46 0.015 [            sftp.Save]   1 sftp.go:397 save <key/5f9df25253>: rename temp-a956b9862f2715b6a8e6: <nil>
2016/08/29 07:16:46 0.000 [Repo.SaveJSONUnpacked]   1 repository.go:250 save new blob config
2016/08/29 07:16:46 0.000 [            sftp.Save]   1 sftp.go:365 save <config/4c5f12becd> bytes to 155
2016/08/29 07:16:46 0.002 [            sftp.Save]   1 sftp.go:379 save <config/4c5f12becd> (155 bytes) to test2/tmp/temp-f3aab2d46a890e29c290
2016/08/29 07:16:46 0.016 [            sftp.Save]   1 sftp.go:397 save <config/4c5f12becd>: rename temp-f3aab2d46a890e29c290: <nil>
2016/08/29 07:16:46 0.019 [Repo.SaveJSONUnpacked]   1 repository.go:277 blob <config/4c5f12becd> saved
2016/08/29 07:16:46 0.000 [            unlockAll]   1 lock.go:111 unlocking 0 locks
2016/08/29 07:16:53 0.000 [    restic]   1 main.go:30 main []string{"./restic", "-r", "sftp:p4:test1", "backup", "/home/wscott/src/restic/README.md"}
2016/08/29 07:16:53 0.000 [      open]   1 global.go:273 parsing location sftp:p4:test1
2016/08/29 07:16:53 0.000 [      open]   1 global.go:284 opening sftp repository at sftp.Config{User:"", Host:"p4", Dir:"test1"}
2016/08/29 07:16:53 0.000 [sftp.OpenWithConfig]   1 sftp.go:154 open with config { p4 test1}
2016/08/29 07:16:53 0.000 [          sftp.Open]   1 sftp.go:118 open backend with program ssh, [p4 -s sftp] at test1
2016/08/29 07:16:55 0.000 [          sftp.List]   1 sftp.go:457 list all key
2016/08/29 07:16:55 0.000 [          SearchKey]   1 key.go:122 trying key 74c95c2bd11e
2016/08/29 07:16:55 0.000 [          sftp.Stat]   1 sftp.go:407 stat <key/74c95c2bd1>
2016/08/29 07:16:55 0.000 [          sftp.Load]   1 sftp.go:328 load <key/74c95c2bd1>, 447 bytes, offset 0
2016/08/29 07:16:55 0.431 [          SearchKey]   1 key.go:135 successfully opened key 74c95c2bd11e
2016/08/29 07:16:55 0.000 [          Repo.Load]   1 repository.go:55 load config with id [null]
2016/08/29 07:16:55 0.431 [          sftp.Stat]   1 sftp.go:407 stat <config/0000000000>
2016/08/29 07:16:55 0.431 [          sftp.Load]   1 sftp.go:328 load <config/0000000000>, 155 bytes, offset 0
2016/08/29 07:16:55 0.000 [          sftp.List]   4 sftp.go:457 list all lock
2016/08/29 07:16:55 0.365 [          Repo.Load]   1 repository.go:55 load lock with id a28e0c23
2016/08/29 07:16:55 0.365 [          sftp.Stat]   1 sftp.go:407 stat <lock/a28e0c235a>
2016/08/29 07:16:55 0.366 [          sftp.Load]   1 sftp.go:328 load <lock/a28e0c235a>, 166 bytes, offset 0
2016/08/29 07:16:55 0.000 [Repo.SaveJSONUnpacked]   1 repository.go:250 save new blob lock
2016/08/29 07:16:55 0.000 [            sftp.Save]   1 sftp.go:365 save <lock/19a3eaee4c> bytes to 166
2016/08/29 07:16:55 0.002 [            sftp.Save]   1 sftp.go:379 save <lock/19a3eaee4c> (166 bytes) to test1/tmp/temp-acb7eee77d42fc194f09
2016/08/29 07:16:55 0.015 [            sftp.Save]   1 sftp.go:397 save <lock/19a3eaee4c>: rename temp-acb7eee77d42fc194f09: <nil>
2016/08/29 07:16:55 0.018 [Repo.SaveJSONUnpacked]   1 repository.go:277 blob <lock/19a3eaee4c> saved
2016/08/29 07:16:56 0.000 [            sftp.List]  35 sftp.go:457 list all lock
2016/08/29 07:16:56 0.270 [            Repo.Load]   1 repository.go:55 load lock with id a28e0c23
2016/08/29 07:16:56 0.270 [            sftp.Stat]   1 sftp.go:407 stat <lock/a28e0c235a>
2016/08/29 07:16:56 0.269 [            sftp.Load]   1 sftp.go:328 load <lock/a28e0c235a>, 166 bytes, offset 0
2016/08/29 07:16:56 0.007 [            Repo.Load]   1 repository.go:55 load lock with id 19a3eaee
2016/08/29 07:16:56 0.007 [            sftp.Stat]   1 sftp.go:407 stat <lock/19a3eaee4c>
2016/08/29 07:16:56 0.007 [            sftp.Load]   1 sftp.go:328 load <lock/19a3eaee4c>, 166 bytes, offset 0
2016/08/29 07:16:56 0.000 [  main.lockRepository]   1 lock.go:42 start goroutine for lock refresh
2016/08/29 07:16:56 0.000 [       Repo.LoadIndex]   1 repository.go:356 Loading index
2016/08/29 07:16:56 0.000 [            sftp.List]   7 sftp.go:457 list all index
2016/08/29 07:16:56 0.000 [    main.refreshLocks]   6 lock.go:58 start
2016/08/29 07:16:56 0.000 [ LoadIndexWithDecoder]  16 index.go:548 Loading index [93 194 28 183 2 96 160 138]
2016/08/29 07:16:56 0.000 [            Repo.Load]  16 repository.go:55 load index with id 5dc21cb7
2016/08/29 07:16:56 0.000 [            sftp.Stat]  16 sftp.go:407 stat <index/5dc21cb702>
2016/08/29 07:16:56 0.000 [            sftp.Load]  16 sftp.go:328 load <index/5dc21cb702>, 155206 bytes, offset 0
2016/08/29 07:16:56 0.000 [    Index.DecodeIndex]  16 index.go:481 Start decoding index
2016/08/29 07:16:56 0.023 [    Index.DecodeIndex]  16 index.go:512 done
2016/08/29 07:16:56 0.000 [            sftp.List]  66 sftp.go:457 list all snapshot
2016/08/29 07:16:56 0.193 [            Repo.Load]   1 repository.go:55 load snapshot with id db6315ab
2016/08/29 07:16:56 0.193 [            sftp.Stat]   1 sftp.go:407 stat <snapshot/db6315abd4>
2016/08/29 07:16:56 0.194 [            sftp.Load]   1 sftp.go:328 load <snapshot/db6315abd4>, 246 bytes, offset 0
2016/08/29 07:16:56 0.000 [                 Scan]   1 archiver.go:782 Start for /home/wscott/src/restic/README.md
2016/08/29 07:16:56 0.000 [                 Scan]   1 archiver.go:820 Done for /home/wscott/src/restic/README.md, err: <nil>
2016/08/29 07:16:56 0.000 [    Archiver.Snapshot]   1 archiver.go:642 start for [/home/wscott/src/restic/README.md]
2016/08/29 07:16:56 0.000 [           pipe.Split]  84 pipe.go:244 start
2016/08/29 07:16:56 0.000 [  ArchivePipe.compare]  83 archiver.go:472 start
2016/08/29 07:16:56 0.000 [  ArchivePipe.compare]  83 archiver.go:485 old channel is closed, copy from new channel
2016/08/29 07:16:56 0.000 [            pipe.Walk]  82 pipe.go:211 start on [/home/wscott/src/restic/README.md]
2016/08/29 07:16:56 0.000 [            pipe.Walk]  82 pipe.go:219 start walker for /home/wscott/src/restic/README.md
2016/08/29 07:16:56 0.000 [            pipe.walk]  82 pipe.go:87 start on "/home/wscott/src/restic/README.md", basedir "/home/wscott/src/restic"
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  96 archiver.go:337 start
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  86 archiver.go:337 start
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  98 archiver.go:337 start
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker] 100 archiver.go:337 start
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  88 archiver.go:337 start
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker] 102 archiver.go:337 start
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker] 104 archiver.go:337 start
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  90 archiver.go:337 start
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  92 archiver.go:337 start
2016/08/29 07:16:56 0.002 [    Archiver.Snapshot]   1 archiver.go:723 wait for workers
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  94 archiver.go:337 start
2016/08/29 07:16:56 0.001 [            pipe.walk]  82 pipe.go:111 sending file job for /home/wscott/src/restic/README.md, res 0xc82020eae0
2016/08/29 07:16:56 0.001 [            pipe.Walk]  82 pipe.go:229 walker for /home/wscott/src/restic/README.md done
2016/08/29 07:16:56 0.000 [            pipe.Walk]  82 pipe.go:232 sending root node, res 0xc82020e780
2016/08/29 07:16:56 0.000 [            pipe.Walk]  82 pipe.go:239 walker done
2016/08/29 07:16:56 0.000 [            pipe.Walk]  82 pipe.go:213 output channel closed
2016/08/29 07:16:56 0.000 [    Archiver.Snapshot]  82 archiver.go:688 pipe.Walk done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  85 archiver.go:263 got job {/home/wscott/src/restic README.md 0xc820212340 <nil> 0xc82020eae0 <nil>}
2016/08/29 07:16:56 0.000 [             copyJobs]  83 archiver.go:448 input channel closed, we're done
2016/08/29 07:16:56 0.002 [  ArchivePipe.compare]  83 archiver.go:469 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  85 archiver.go:306    README.md no old data
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  85 archiver.go:311    read and save README.md, content: []
2016/08/29 07:16:56 0.002 [   Archiver.dirWorker]  96 archiver.go:349 save dir  (1 entries), error <nil>
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  96 archiver.go:363 receiving result from 0xc82020eae0
2016/08/29 07:16:56 0.003 [           pipe.Split]  84 pipe.go:264 done
2016/08/29 07:16:56 0.000 [    Archiver.Snapshot]  84 archiver.go:703 split done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  93 archiver.go:252 done
2016/08/29 07:16:56 0.002 [   Archiver.dirWorker]  90 archiver.go:339 done
2016/08/29 07:16:56 0.003 [   Archiver.dirWorker]  86 archiver.go:339 done
2016/08/29 07:16:56 0.003 [   Archiver.dirWorker]  98 archiver.go:339 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  87 archiver.go:252 done
2016/08/29 07:16:56 0.003 [   Archiver.dirWorker] 100 archiver.go:339 done
2016/08/29 07:16:56 0.002 [   Archiver.dirWorker]  92 archiver.go:339 done
2016/08/29 07:16:56 0.003 [   Archiver.dirWorker]  88 archiver.go:339 done
2016/08/29 07:16:56 0.002 [   Archiver.dirWorker]  94 archiver.go:339 done
2016/08/29 07:16:56 0.003 [   Archiver.dirWorker] 102 archiver.go:339 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  95 archiver.go:252 done
2016/08/29 07:16:56 0.003 [   Archiver.dirWorker] 104 archiver.go:339 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  97 archiver.go:252 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  89 archiver.go:252 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  99 archiver.go:252 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker] 103 archiver.go:252 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker] 101 archiver.go:252 done
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  91 archiver.go:252 done
2016/08/29 07:16:56 0.000 [        Archiver.Save]  68 archiver.go:95 Save(data, 145dde52)
2016/08/29 07:16:56 0.000 [   MasterIndex.Lookup]  68 master_index.go:28 looking up id 145dde52, tpe data
2016/08/29 07:16:56 0.000 [         Index.Lookup]  68 index.go:125 id 145dde52 found in pack 3adcf87a at 1888144, length 3653
2016/08/29 07:16:56 0.000 [   MasterIndex.Lookup]  68 master_index.go:34 found id 145dde52: [<PackedBlob 145dde52 type data in pack 3adcf87a: len 3653, offset 1888144]
2016/08/29 07:16:56 0.001 [        Archiver.Save]  68 archiver.go:98 blob 145dde52 is known
2016/08/29 07:16:56 0.000 [        Archiver.Save]  85 archiver.go:188 checking size for file /home/wscott/src/restic/README.md
2016/08/29 07:16:56 0.000 [        Archiver.Save]  85 archiver.go:197   adding blob 145dde52, 3621 bytes
2016/08/29 07:16:56 0.000 [    Archiver.SaveFile]  85 archiver.go:204 SaveFile("/home/wscott/src/restic/README.md"): 1 blobs
2016/08/29 07:16:56 0.012 [  Archiver.fileWorker]  85 archiver.go:326    processed README.md, 1/0 blobs
2016/08/29 07:16:56 0.000 [  Archiver.fileWorker]  85 archiver.go:252 done
2016/08/29 07:16:56 0.000 [   MasterIndex.Lookup]  96 master_index.go:28 looking up id e4a8cc4d, tpe tree
2016/08/29 07:16:56 0.000 [         Index.Lookup]  96 index.go:141 id e4a8cc4d not found
2016/08/29 07:16:56 0.000 [   MasterIndex.Lookup]  96 master_index.go:39 id e4a8cc4d not found in any index
2016/08/29 07:16:56 0.000 [        Repo.SaveJSON]  96 repository.go:231 save tree blob
2016/08/29 07:16:56 0.000 [            Repo.Save]  96 repository.go:192 save id e4a8cc4d (tree, 351 bytes)
2016/08/29 07:16:56 0.000 [      Repo.findPacker]  96 packer_manager.go:70 create new pack for 383 bytes
2016/08/29 07:16:56 0.001 [            Repo.Save]  96 repository.go:219 pack is not full enough (383 bytes)
2016/08/29 07:16:56 0.000 [    Repo.insertPacker]  96 packer_manager.go:85 1 packers
2016/08/29 07:16:56 0.015 [   Archiver.dirWorker]  96 archiver.go:406 save tree for : e4a8cc4d
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  96 archiver.go:413 sending result to 0xc82020e780
2016/08/29 07:16:56 0.000 [   Archiver.dirWorker]  96 archiver.go:339 done
2016/08/29 07:16:56 0.017 [    Archiver.Snapshot]   1 archiver.go:730 workers terminated
2016/08/29 07:16:56 0.000 [    Archiver.Snapshot]   1 archiver.go:734 root node received: e4a8cc4d
2016/08/29 07:16:56 0.473 [Repo.SaveJSONUnpacked]   1 repository.go:250 save new blob snapshot
2016/08/29 07:16:56 0.473 [            sftp.Save]   1 sftp.go:365 save <snapshot/e1239daad2> bytes to 255
2016/08/29 07:16:56 0.002 [            sftp.Save]   1 sftp.go:379 save <snapshot/e1239daad2> (255 bytes) to test1/tmp/temp-03e316ca4115c9d43c86
2016/08/29 07:16:56 0.018 [            sftp.Save]   1 sftp.go:397 save <snapshot/e1239daad2>: rename temp-03e316ca4115c9d43c86: <nil>
2016/08/29 07:16:56 0.020 [Repo.SaveJSONUnpacked]   1 repository.go:277 blob <snapshot/e1239daad2> saved
2016/08/29 07:16:56 0.020 [    Archiver.Snapshot]   1 archiver.go:745 saved snapshot e1239daa
2016/08/29 07:16:56 0.000 [           Repo.Flush]   1 repository.go:286 manually flushing 1 packs
2016/08/29 07:16:56 0.000 [      Repo.savePacker]   1 packer_manager.go:90 save packer with 1 blobs
2016/08/29 07:16:56 0.001 [            sftp.Save]   1 sftp.go:365 save <data/c37bad7cb1> bytes to 456
2016/08/29 07:16:56 0.002 [            sftp.Save]   1 sftp.go:379 save <data/c37bad7cb1> (456 bytes) to test1/tmp/temp-c57ec342adee0a1f0bca
2016/08/29 07:16:56 0.027 [            sftp.Save]   1 sftp.go:397 save <data/c37bad7cb1>: rename temp-c57ec342adee0a1f0bca: <nil>
2016/08/29 07:16:56 0.029 [      Repo.savePacker]   1 packer_manager.go:122 saved as <data/c37bad7cb1>
2016/08/29 07:16:56 0.000 [      Repo.savePacker]   1 packer_manager.go:131   updating blob e4a8cc4d to pack c37bad7c
2016/08/29 07:16:56 0.000 [          Index.Store]   1 index.go:108 <PackedBlob e4a8cc4d type tree in pack c37bad7c: len 383, offset 0
2016/08/29 07:16:56 0.000 [MasterIndex.NotFinalIndexes]   1 master_index.go:158 return 1 indexes
2016/08/29 07:16:56 0.000 [             Repo.SaveIndex]   1 repository.go:328 Saving index 0
2016/08/29 07:16:56 0.000 [               Index.Encode]   1 index.go:393 encoding index
2016/08/29 07:16:56 0.000 [               Index.encode]   1 index.go:376 encoding index
2016/08/29 07:16:56 0.000 [     Index.generatePackList]   1 index.go:324 handle blob <tree/e4a8cc4d>
2016/08/29 07:16:56 0.000 [     Index.generatePackList]   1 index.go:353 done
2016/08/29 07:16:56 0.001 [                  sftp.Save]   1 sftp.go:365 save <index/5da850ddf0> bytes to 240
2016/08/29 07:16:56 0.002 [                  sftp.Save]   1 sftp.go:379 save <index/5da850ddf0> (240 bytes) to test1/tmp/temp-c925b712fae8d3583717
2016/08/29 07:16:56 0.016 [                  sftp.Save]   1 sftp.go:397 save <index/5da850ddf0>: rename temp-c925b712fae8d3583717: <nil>
2016/08/29 07:16:56 0.049 [      Repo.SaveJSONUnpacked]   1 repository.go:277 blob <index/5da850ddf0> saved
2016/08/29 07:16:56 0.019 [             Repo.SaveIndex]   1 repository.go:335 Saved index 0 as 5da850dd
2016/08/29 07:16:56 0.049 [          Archiver.Snapshot]   1 archiver.go:760 saved indexes
2016/08/29 07:16:56 0.000 [                 unlockRepo]   1 lock.go:91 unlocking repository
2016/08/29 07:16:56 0.000 [                sftp.Remove]   1 sftp.go:445 type lock, name 19a3eaee4c5254fa0cb68aa4c1d1a4f972b7ff0b606f8265a1ae4cb06b8d2d08
2016/08/29 07:16:56 0.004 [                 unlockRepo]   1 lock.go:93 error while unlocking: file does not exist
2016/08/29 07:16:56 0.000 [                  unlockAll]   1 lock.go:111 unlocking 1 locks
2016/08/29 07:16:56 0.004 [                sftp.Remove]   1 sftp.go:445 type lock, name 19a3eaee4c5254fa0cb68aa4c1d1a4f972b7ff0b606f8265a1ae4cb06b8d2d08
2016/08/29 07:16:56 0.003 [                  unlockAll]   1 lock.go:114 error while unlocking: file does not exist
```

'no data' was with the debug log crash.

It does appear to be writing the data:
```
$ DEBUG_LOG=/tmp/restic-debug.log ./restic -r sftp:p4:test1 snapshots
debug log file /tmp/restic-debug.log
debug enabled
enter password for repository: 
ID        Date                 Host        Directory
----------------------------------------------------------------------
db6315ab  2016-08-29 07:15:42  x1          /home/wscott/src/restic
e1239daa  2016-08-29 07:16:56  x1          /home/wscott/src/restic/README.md
error in cleanup handler: file does not exist
```

log:
```
2016/08/29 07:57:29 0.000 [    restic]   1 main.go:30 main []string{"./restic", "-r", "sftp:p4:test1", "sn
2016/08/29 07:57:29 0.000 [      open]   1 global.go:273 parsing location sftp:p4:test1
2016/08/29 07:57:29 0.000 [      open]   1 global.go:284 opening sftp repository at sftp.Config{User:"", H
2016/08/29 07:57:29 0.000 [sftp.OpenWithConfig]   1 sftp.go:154 open with config { p4 test1}
2016/08/29 07:57:29 0.000 [          sftp.Open]   1 sftp.go:118 open backend with program ssh, [p4 -s sftp
2016/08/29 07:57:31 0.000 [          sftp.List]   1 sftp.go:457 list all key
2016/08/29 07:57:31 0.000 [          SearchKey]   1 key.go:122 trying key 74c95c2bd11e
2016/08/29 07:57:31 0.000 [          sftp.Stat]   1 sftp.go:407 stat <key/74c95c2bd1>
2016/08/29 07:57:31 0.000 [          sftp.Load]   1 sftp.go:328 load <key/74c95c2bd1>, 447 bytes, offset 0
2016/08/29 07:57:32 0.409 [          SearchKey]   1 key.go:135 successfully opened key 74c95c2bd11e
2016/08/29 07:57:32 0.000 [          Repo.Load]   1 repository.go:55 load config with id [null]
2016/08/29 07:57:32 0.408 [          sftp.Stat]   1 sftp.go:407 stat <config/0000000000>
2016/08/29 07:57:32 0.409 [          sftp.Load]   1 sftp.go:328 load <config/0000000000>, 155 bytes, offse
2016/08/29 07:57:32 0.000 [          sftp.List]  50 sftp.go:457 list all lock
2016/08/29 07:57:32 0.359 [          Repo.Load]   1 repository.go:55 load lock with id a28e0c23
2016/08/29 07:57:32 0.360 [          sftp.Stat]   1 sftp.go:407 stat <lock/a28e0c235a>
2016/08/29 07:57:32 0.359 [          sftp.Load]   1 sftp.go:328 load <lock/a28e0c235a>, 166 bytes, offset 
2016/08/29 07:57:32 0.007 [          Repo.Load]   1 repository.go:55 load lock with id 19a3eaee
2016/08/29 07:57:32 0.007 [          sftp.Stat]   1 sftp.go:407 stat <lock/19a3eaee4c>
2016/08/29 07:57:32 0.007 [          sftp.Load]   1 sftp.go:328 load <lock/19a3eaee4c>, 166 bytes, offset 
0
2016/08/29 07:57:32 0.000 [Repo.SaveJSONUnpacked]   1 repository.go:250 save new blob lock
2016/08/29 07:57:32 0.000 [            sftp.Save]   1 sftp.go:365 save <lock/48b6a80097> bytes to 166
2016/08/29 07:57:32 0.002 [            sftp.Save]   1 sftp.go:379 save <lock/48b6a80097> (166 bytes) to te
2016/08/29 07:57:32 0.018 [            sftp.Save]   1 sftp.go:397 save <lock/48b6a80097>: rename temp-210d
2016/08/29 07:57:32 0.021 [Repo.SaveJSONUnpacked]   1 repository.go:277 blob <lock/48b6a80097> saved
2016/08/29 07:57:32 0.000 [            sftp.List]  52 sftp.go:457 list all lock
2016/08/29 07:57:32 0.269 [            Repo.Load]   1 repository.go:55 load lock with id a28e0c23
2016/08/29 07:57:32 0.269 [            sftp.Stat]   1 sftp.go:407 stat <lock/a28e0c235a>
2016/08/29 07:57:32 0.269 [            sftp.Load]   1 sftp.go:328 load <lock/a28e0c235a>, 166 bytes, offse
2016/08/29 07:57:32 0.007 [            Repo.Load]   1 repository.go:55 load lock with id 19a3eaee
2016/08/29 07:57:32 0.007 [            sftp.Stat]   1 sftp.go:407 stat <lock/19a3eaee4c>
2016/08/29 07:57:32 0.007 [            sftp.Load]   1 sftp.go:328 load <lock/19a3eaee4c>, 166 bytes, offse
2016/08/29 07:57:32 0.007 [            Repo.Load]   1 repository.go:55 load lock with id 48b6a800
2016/08/29 07:57:32 0.007 [            sftp.Stat]   1 sftp.go:407 stat <lock/48b6a80097>
2016/08/29 07:57:32 0.007 [            sftp.Load]   1 sftp.go:328 load <lock/48b6a80097>, 166 bytes, offse
2016/08/29 07:57:32 0.000 [  main.lockRepository]   1 lock.go:42 start goroutine for lock refresh
2016/08/29 07:57:32 0.000 [            sftp.List]  21 sftp.go:457 list all snapshot
2016/08/29 07:57:32 0.000 [    main.refreshLocks]  20 lock.go:58 start
2016/08/29 07:57:32 0.055 [            Repo.Load]   1 repository.go:55 load snapshot with id db6315ab
2016/08/29 07:57:32 0.055 [            sftp.Stat]   1 sftp.go:407 stat <snapshot/db6315abd4>
2016/08/29 07:57:32 0.055 [            sftp.Load]   1 sftp.go:328 load <snapshot/db6315abd4>, 246 bytes, o
2016/08/29 07:57:32 0.007 [            Repo.Load]   1 repository.go:55 load snapshot with id e1239daa
2016/08/29 07:57:32 0.007 [            sftp.Stat]   1 sftp.go:407 stat <snapshot/e1239daad2>
2016/08/29 07:57:32 0.007 [            sftp.Load]   1 sftp.go:328 load <snapshot/e1239daad2>, 255 bytes, o
2016/08/29 07:57:32 1.178 [            sftp.List]   1 sftp.go:457 list all snapshot
2016/08/29 07:57:33 0.000 [           unlockRepo]   1 lock.go:91 unlocking repository
2016/08/29 07:57:33 0.000 [          sftp.Remove]   1 sftp.go:445 type lock, name 48b6a8009757f9f8e163c7c2
2016/08/29 07:57:33 0.003 [           unlockRepo]   1 lock.go:93 error while unlocking: file does not exis
2016/08/29 07:57:33 0.000 [            unlockAll]   1 lock.go:111 unlocking 1 locks
2016/08/29 07:57:33 0.002 [          sftp.Remove]   1 sftp.go:445 type lock, name 48b6a8009757f9f8e163c7c2
2016/08/29 07:57:33 0.002 [            unlockAll]   1 lock.go:114 error while unlocking: file does not exi
```

So the title of this issue is now wrong.

On my lock name question, I was looking at a different output than the one I pasted. In the paste it is

  sftp.go:365 save lock/48b6a80097 bytes to 166

at the top and this at the bottom

   sftp.go:445 type lock, name 48b6a8009757f9f8e163c7c266cd1cfdd3242ac755b9da29a890d2e4f165b023
  lock.go:114 error while unlocking: file does not exist

Kinda looks like it is deleting a different filename than it created above.

For info, here's what happens with restic 0.8 on an up to date synology DS212j (DSM 6.1.4-15217 Update 3). Message are not exactly the same as above, but the fix is the same, so most likely the same issue. Adding this so that searching for the error message also sends people here.

```
$ restic -r sftp:user@192.168.1.2:marc/backup-restic init
enter password for new backend:
enter password again:
created restic backend 9334198711 at sftp:user@192.168.1.2:marc/backup-restic

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
$ restic -r sftp:marc@192.168.1.2:marc/backup-restic backup photos/
enter password for repository:
password is correct
scan [/mnt/foo/bar/photos]
scanned 1563 directories, 151916 files in 0:03
Save(<data/7524a2cd7e>) returned error, retrying after 747.292382ms: OpenFile: EOF
Save(<data/aa9850380c>) returned error, retrying after 570.255626ms: cmd.Wait: exit status 255
Save(<data/59ce5f44c8>) returned error, retrying after 374.63265ms: OpenFile: failed to send packet header: write |1: file already closed
Save(<data/b5ca2a6519>) returned error, retrying after 443.757345ms: OpenFile: failed to send packet header: write |1:file already closed
... [lots of these]
[0:43] 0.01%  927.827 KiB/s  38.961 MiB / 445.802 GiB  125 / 153479 items  0 errors  ETA 139:56:16
error while saving data to the repo: failed to send packet header: write |1: file already closed
github.com/pkg/sftp.sendPacket
	github.com/pkg/sftp/packet.go:132
github.com/pkg/sftp.(*conn).sendPacket
	github.com/pkg/sftp/conn.go:31
github.com/pkg/sftp.(*clientConn).dispatchRequest
	github.com/pkg/sftp/conn.go:106
github.com/pkg/sftp.(*clientConn).sendPacket
	github.com/pkg/sftp/conn.go:97
github.com/pkg/sftp.(*Client).open
	github.com/pkg/sftp/client.go:381
github.com/pkg/sftp.(*Client).OpenFile
	github.com/pkg/sftp/client.go:376
github.com/restic/restic/internal/backend/sftp.(*SFTP).Save
	github.com/restic/restic/internal/backend/sftp/sftp.go:336
github.com/restic/restic/internal/backend.(*RetryBackend).Save.func1
	github.com/restic/restic/internal/backend/backend_retry.go:68
github.com/cenkalti/backoff.RetryNotify
	github.com/cenkalti/backoff/retry.go:37
github.com/restic/restic/internal/backend.(*RetryBackend).retry
	github.com/restic/restic/internal/backend/backend_retry.go:36
github.com/restic/restic/internal/backend.(*RetryBackend).Save
	github.com/restic/restic/internal/backend/backend_retry.go:62
github.com/restic/restic/internal/cache.(*Backend).Save
	github.com/restic/restic/internal/cache/backend.go:55
github.com/restic/restic/internal/repository.(*Repository).savePacker
	github.com/restic/restic/internal/repository/packer_manager.go:107
github.com/restic/restic/internal/repository.(*Repository).SaveAndEncrypt
	github.com/restic/restic/internal/repository/repository.go:253
github.com/restic/restic/internal/repository.(*Repository).SaveBlob
	github.com/restic/restic/internal/repository/repository.go:618
github.com/restic/restic/internal/archiver.(*Archiver).Save
	github.com/restic/restic/internal/archiver/archiver.go:104
github.com/restic/restic/internal/archiver.(*Archiver).saveChunk
	github.com/restic/restic/internal/archiver/archiver.go:161
runtime.goexit
	/usr/lib/go-1.9/src/runtime/asm_amd64.s:2337
OpenFile
github.com/restic/restic/internal/backend/sftp.(*SFTP).Save
	github.com/restic/restic/internal/backend/sftp/sftp.go:348
github.com/restic/restic/internal/backend.(*RetryBackend).Save.func1
	github.com/restic/restic/internal/backend/backend_retry.go:68
github.com/cenkalti/backoff.RetryNotify
	github.com/cenkalti/backoff/retry.go:37
github.com/restic/restic/internal/backend.(*RetryBackend).retry
	github.com/restic/restic/internal/backend/backend_retry.go:36
github.com/restic/restic/internal/backend.(*RetryBackend).Save
	github.com/restic/restic/internal/backend/backend_retry.go:62
github.com/restic/restic/internal/cache.(*Backend).Save
	github.com/restic/restic/internal/cache/backend.go:55
github.com/restic/restic/internal/repository.(*Repository).savePacker
	github.com/restic/restic/internal/repository/packer_manager.go:107
github.com/restic/restic/internal/repository.(*Repository).SaveAndEncrypt
	github.com/restic/restic/internal/repository/repository.go:253
github.com/restic/restic/internal/repository.(*Repository).SaveBlob
	github.com/restic/restic/internal/repository/repository.go:618
github.com/restic/restic/internal/archiver.(*Archiver).Save
	github.com/restic/restic/internal/archiver/archiver.go:104
github.com/restic/restic/internal/archiver.(*Archiver).saveChunk
	github.com/restic/restic/internal/archiver/archiver.go:161
runtime.goexit
	/usr/lib/go-1.9/src/runtime/asm_amd64.s:2337
panic: OpenFile: failed to send packet header: write |1: file already closed

goroutine 206 [running]:
github.com/restic/restic/internal/archiver.(*Archiver).saveChunk(0xc42018b0e0, 0xe0c8e0, 0xc420018100, 0x0, 0x55800, 0x
1, 0xc420ed0000, 0x55800, 0x80000, 0xc4201920b0, ...)
	github.com/restic/restic/internal/archiver/archiver.go:166 +0x41b
created by github.com/restic/restic/internal/archiver.(*Archiver).SaveFile
	github.com/restic/restic/internal/archiver/archiver.go:240 +0x3ff
```
Using absolute path fixes the issue.

--- a/changelog/unreleased/issue-2518
+++ b/changelog/unreleased/issue-2518
@@ -1,0 +1,16 @@
+Bugfix: Do not crash with Synology NAS sftp server
+
+It was found that when restic is used to store data on an sftp server on a
+Synology NAS with a relative path (one which does not start with a slash), it
+may go into an endless loop trying to create directories on the server. We've
+fixed this bug by using a function oft the sftp library instead of our own
+implementation.
+
+The bug was discovered because the Synology sftp server behaves erratic with
+non-absolute path (e.g. `home/restic-repo`). This can be resolved by just using
+an absolute path instead (`/home/restic-repo`). We've also added a paragraph in
+the FAQ.
+
+https://github.com/restic/restic/issues/2518
+https://github.com/restic/restic/issues/2363
+https://github.com/restic/restic/pull/2530

--- a/changelog/unreleased/issue-2518
+++ b/changelog/unreleased/issue-2518
@@ -3,7 +3,7 @@ Bugfix: Do not crash with Synology NAS sftp server
 It was found that when restic is used to store data on an sftp server on a
 Synology NAS with a relative path (one which does not start with a slash), it
 may go into an endless loop trying to create directories on the server. We've
-fixed this bug by using a function oft the sftp library instead of our own
+fixed this bug by using a function in the sftp library instead of our own
 implementation.
 
 The bug was discovered because the Synology sftp server behaves erratic with

--- a/doc/faq.rst
+++ b/doc/faq.rst
@@ -142,6 +142,9 @@ is not slowed down, which is particularly useful for servers.
 Creating new repo on a Synology NAS via sftp fails
 --------------------------------------------------
 
+For using restic with a Synology NAS via sftp, please make sure that the
+specified path is absolute, it must start with a slash (``/``).
+
 Sometimes creating a new restic repository on a Synology NAS via sftp fails
 with an error similar to the following:
 
@@ -160,8 +163,8 @@ different than the directory structure on the device and maybe even as exposed
 via other protocols.
 
 
-Try removing the /volume1 prefix in your paths. If this does not work, use sftp
-and ls to explore the SFTP file system hierarchy on your NAS.
+Try removing the ``/volume1`` prefix in your paths. If this does not work, use
+sftp and ls to explore the SFTP file system hierarchy on your NAS.
 
 The following may work:
 

--- a/internal/backend/sftp/sftp.go
+++ b/internal/backend/sftp/sftp.go
@@ -138,8 +138,7 @@ func Open(cfg Config) (*SFTP, error) {
 
 func (r *SFTP) mkdirAllDataSubdirs() error {
 	for _, d := range r.Paths() {
-		err := r.mkdirAll(d, backend.Modes.Dir)
-		debug.Log("mkdirAll %v -> %v", d, err)
+		err := r.c.MkdirAll(d)
 		if err != nil {
 			return err
 		}
@@ -250,38 +249,6 @@ func (r *SFTP) Location() string {
 	return r.p
 }
 
-func (r *SFTP) mkdirAll(dir string, mode os.FileMode) error {
-	// check if directory already exists
-	fi, err := r.c.Lstat(dir)
-	if err == nil {
-		if fi.IsDir() {
-			return nil
-		}
-
-		return errors.Errorf("mkdirAll(%s): entry exists but is not a directory", dir)
-	}
-
-	// create parent directories
-	errMkdirAll := r.mkdirAll(path.Dir(dir), backend.Modes.Dir)
-
-	// create directory
-	errMkdir := r.c.Mkdir(dir)
-
-	// test if directory was created successfully
-	fi, err = r.c.Lstat(dir)
-	if err != nil {
-		// return previous errors
-		return errors.Errorf("mkdirAll(%s): unable to create directories: %v, %v", dir, errMkdirAll, errMkdir)
-	}
-
-	if !fi.IsDir() {
-		return errors.Errorf("mkdirAll(%s): entry exists but is not a directory", dir)
-	}
-
-	// set mode
-	return r.c.Chmod(dir, mode)
-}
-
 // Join joins the given paths and cleans them afterwards. This always uses
 // forward slashes, which is required by sftp.
 func Join(parts ...string) string {
@@ -306,7 +273,7 @@ func (r *SFTP) Save(ctx context.Context, h restic.Handle, rd restic.RewindReader
 
 	if r.IsNotExist(err) {
 		// error is caused by a missing directory, try to create it
-		mkdirErr := r.mkdirAll(r.Dirname(h), backend.Modes.Dir)
+		mkdirErr := r.c.MkdirAll(r.Dirname(h))
 		if mkdirErr != nil {
 			debug.Log("error creating dir %v: %v", r.Dirname(h), mkdirErr)
 		} else {

